File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/ui/DisplayArgsLoader.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.ui;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

/**
 * A display args loader will store a reference to the
 * layout args and restore the instance when it to when
 * loading from a parcel without actually placing the args
 * in a parcel.
 *
 * This only works in a single process.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class DisplayArgsLoader implements Parcelable {
    private final static Map<String, DisplayArgs> cached = new HashMap<>();

    private final String id;

    private DisplayArgsLoader(@Nullable String id) {
        this.id = id;
    }

    private DisplayArgsLoader(Parcel in) {
        this.id = in.readString();
    }

    public static final Creator<DisplayArgsLoader> CREATOR = new Creator<DisplayArgsLoader>() {
        @Override
        public DisplayArgsLoader createFromParcel(Parcel in) {
            return new DisplayArgsLoader(in);
        }

        @Override
        public DisplayArgsLoader[] newArray(int size) {
            return new DisplayArgsLoader[size];
        }
    };

    /**
     * Creates a new layout args loader. The args will be cached, to remove them from
     * the cache call `dispose` on the instance.
     *
     * @param args The layout args.
     * @return A layout args loader.
     */
    public static DisplayArgsLoader newLoader(@NonNull DisplayArgs args) {
        String loaderId = UUID.randomUUID().toString();
        cached.put(loaderId, args);
        return new DisplayArgsLoader(loaderId);
    }

    public void dispose() {
        cached.remove(id);
    }

    @NonNull
    public DisplayArgs getLayoutArgs() throws LoadException {
        DisplayArgs args = cached.get(id);
        if (args == null) {
            throw new LoadException("Layout args no longer available");
        }

        return args;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
    }

    public static class LoadException extends Exception {
        public LoadException(String message) {
            super(message);
        }
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/display/DisplayArgsLoader.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.display

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.RestrictTo
import java.util.UUID

/**
 * A display args loader will store a reference to the
 * layout args and restore the instance when it to when
 * loading from a parcel without actually placing the args
 * in a parcel.
 *
 * This only works in a single process.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class DisplayArgsLoader : Parcelable {

    private val id: String?

    private constructor(id: String?) {
        this.id = id
    }

    private constructor(parcel: Parcel) {
        this.id = parcel.readString()
    }

    public fun dispose() {
        cached.remove(id)
    }

    @Throws(LoadException::class)
    public fun getDisplayArgs(): DisplayArgs {
        return cached[id] ?: run {
            throw LoadException("Layout args no longer available")
        }
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
    }

    public companion object {

        private val cached = mutableMapOf<String, DisplayArgs>()

        @JvmField
        public val CREATOR: Parcelable.Creator<DisplayArgsLoader> =
            object : Parcelable.Creator<DisplayArgsLoader> {
                override fun createFromParcel(`in`: Parcel): DisplayArgsLoader {
                    return DisplayArgsLoader(`in`)
                }

                override fun newArray(size: Int): Array<DisplayArgsLoader> {
                    return ArrayList<DisplayArgsLoader>(size).toTypedArray()
                }

            }

        /**
         * Creates a new layout args loader. The args will be cached, to remove them from
         * the cache call `dispose` on the instance.
         *
         * @param args The layout args.
         * @return A layout args loader.
         */
        public fun newLoader(args: DisplayArgs): DisplayArgsLoader {
            val loaderId = UUID.randomUUID().toString()
            cached.put(loaderId, args)
            return DisplayArgsLoader(loaderId)
        }
    }

    public class LoadException public constructor(message: String) : Exception(message)
}
